flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/InputProcessorUtilTest.java

```diff
@@ -58,4 +79,57 @@ public void testGenerateInputGateToChannelIndexOffsetMap() {
 		assertEquals(0, inputGateToChannelIndexOffsetMap.get(ig1).intValue());
 		assertEquals(3, inputGateToChannelIndexOffsetMap.get(ig2).intValue());
 	}
+
+	@Test
+	public void testCreateCheckpointedMultipleInputGate() throws Exception {
+		try (CloseableRegistry registry = new CloseableRegistry()) {
+			MockEnvironment environment = new MockEnvironmentBuilder().build();
+			MockStreamTask streamTask = new MockStreamTaskBuilder(environment).build();
+			StreamConfig streamConfig = new StreamConfig(environment.getJobConfiguration());
+			streamConfig.setCheckpointMode(CheckpointingMode.EXACTLY_ONCE);
+			streamConfig.setUnalignedCheckpointsEnabled(true);
+
+			// First input gate has index larger than the second
+			Collection<IndexedInputGate>[] inputGates = new Collection[] {
+				Collections.singletonList(new MockIndexedInputGate(1, 4)),
+				Collections.singletonList(new MockIndexedInputGate(0, 2)),
+			};
+
+			new MockChannelStateWriter() {
+				@Override
+				public void addInputData(
+					long checkpointId,
+					InputChannelInfo info,
+					int startSeqNum,
+					Buffer... data) {
+					super.addInputData(checkpointId, info, startSeqNum, data);
+				}
+			};
+
+			CheckpointedInputGate[] checkpointedMultipleInputGate = InputProcessorUtil.createCheckpointedMultipleInputGate(
+				streamTask,
+				streamConfig,
+				new MockChannelStateWriter(),
+				environment.getMetricGroup().getIOMetricGroup(),
+				streamTask.getName(),
+				inputGates);
+			for (CheckpointedInputGate checkpointedInputGate : checkpointedMultipleInputGate) {
+				registry.registerCloseable(checkpointedInputGate);
+			}
+
+			CheckpointBarrierHandler barrierHandler = checkpointedMultipleInputGate[0].getCheckpointBarrierHandler();
+
+			List<IndexedInputGate> allInputGates = Arrays.stream(inputGates).flatMap(gates -> gates.stream()).collect(Collectors.toList());
+			for (IndexedInputGate inputGate : allInputGates) {
+				for (int channelId = 0; channelId < inputGate.getNumberOfInputChannels(); channelId++) {
+					assertTrue(barrierHandler.getBufferReceivedListener().isPresent());
+					BufferReceivedListener bufferReceivedListener = barrierHandler.getBufferReceivedListener().get();
+					bufferReceivedListener.notifyBarrierReceived(
+						new CheckpointBarrier(1, 42, CheckpointOptions.forCheckpointWithDefaultLocation(true, true)),
+						new InputChannelInfo(inputGate.getGateIndex(), channelId));
+				}
+			}
+			assertFalse(barrierHandler.isCheckpointPending());
```

Comment:
I am not sure why we need to verify this because it never process the barrier in this test. Maybe verify `assertTrue(barrierHandler.getAllBarriersReceivedFuture(1).isDone())` instead?